src/mpi/coll/op/opequal.c

```diff
@@ -0,0 +1,241 @@
+    long double value;
+    int is_equal;
+} MPIR_longdoubleint_eqltype;
+#endif
+
+#define MPIR_EQUAL_FLOAT_PRECISION 1e-6
+
+#define MPIR_EQUAL_FLOAT_COMPARE(_aValue, _bValue)              \
+    (((_aValue <= _bValue + MPIR_EQUAL_FLOAT_PRECISION)         \
+    && (_aValue >= _bValue - MPIR_EQUAL_FLOAT_PRECISION))?      \
+    1:0)
+
+/* If a child found unequal, its parent sticks to unequal. */
+/* Values of is_equal: 1 equal 0 not equal, init using any value other than 0.*/
+#define MPIR_EQUAL_C_CASE_INT(c_type_) {                \
+        c_type_ *a = (c_type_ *)inoutvec;               \
+        c_type_ *b = (c_type_ *)invec;                  \
+        for (i = 0; i < len; ++i) {                         \
+            if (0 == b[i].is_equal || 0 == a[i].is_equal){    \
+                a[i].is_equal = 0;            			\
+            }else if (a[i].value != b[i].value){        \
+                a[i].is_equal = 0;                       \
+            }else{                                      \
+                a[i].is_equal = 1;                       \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+#define MPIR_EQUAL_C_CASE_FLOAT(c_type_) {              \
+        c_type_ *a = (c_type_ *)inoutvec;               \
+        c_type_ *b = (c_type_ *)invec;                  \
+        for (i = 0; i < len; ++i) {                         \
+            if (0 == b[i].is_equal || 0 == a[i].is_equal){    \
+                a[i].is_equal = 0;            			\
+            }else if (!MPIR_EQUAL_FLOAT_COMPARE(a[i].value, b[i].value)){        \
+                a[i].is_equal = 0;                       \
+            }else{                                      \
+                a[i].is_equal = 1;                       \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+#define MPIR_EQUAL_F_CASE(f_type_) {                    \
+        f_type_ *a = (f_type_ *)inoutvec;               \
+        f_type_ *b = (f_type_ *)invec;                  \
+        for (i = 0; i < flen; i += 2) {                       \
+            if(0 == b[i+1] || 0 == a[i+1]){             \
+                a[i+1] = 0;                        		\
+            }else if (a[i] != b[i]){                    \
+                a[i+1] = 0;                             \
+            }else{                                      \
+                a[i+1] = 1;                             \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+void MPIR_EQUAL_user_defined_datatype_compare(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
```

Comment:
Rename to `MPIR_EQUAL_struct`. The `struct` is expected to be constructed by `MPIR_Allreduce_equal`, not by user. And we only expect the `struct` have 2-types, one is a basic integer type with count, and the second is a single integer as `is_equal`. That should keep the implementation simple.

Reply:
Renamed